contracts/contracts/staking/SingleAssetStaking.sol

```diff
@@ -0,0 +1,412 @@
+                        )
+                    )
+                );
+            }
+        }
+    }
+
+    /* ========== MUTATIVE FUNCTIONS ========== */
+
+    /**
+     * @dev Make a preapproved stake for the user, this is a presigned voucher that the user can redeem either from
+     *      an airdrop or a compensation program.
+     *      Only 1 of each type is allowed per user. Signature must be done by the preApprover.
+     * @param stakeType Number that represent the type of the stake, must not be 0 which is user stake
+     * @param duration Number of seconds this stake will be held for
+     * @param rate Rate(0.3 is 30%) of reward for this stake in 1e18, uint240 to fit the bool and type in struct Stake
+     * @param amount Number of tokens to stake in 1e18
+     * @param v Signature v component
+     * @param r Signature r component
+     * @param s Signature s component
+     */
+    function preApprovedStake(
+        uint8 stakeType,
+        uint256 duration,
+        uint256 rate,
+        uint256 amount,
+        uint8 v,
+        bytes32 r,
+        bytes32 s
+    ) external {
+        require(stakeType != USER_STAKE_TYPE, "Cannot be normal staking");
+
+        // message length should be 117 because (uint8)1 + (address) 20 + (uint256)32 + (uint256)32 + (uint256)32
+        bytes32 messageDigest = keccak256(
+            abi.encodePacked(
+                "\x19Ethereum Signed Message:\n117",
+                stakeType,
+                address(msg.sender),
+                duration,
+                rate,
+                amount
+            )
+        );
+
+        require(
+            preApprover == ecrecover(messageDigest, v, r, s),
+            "Stake not approved"
+        );
+
+        // verify that we haven't already staked
+        Stake[] storage stakes = userStakes[msg.sender];
+        for (uint256 i = 0; i < stakes.length; i++) {
+            require(stakes[i].stakeType != stakeType, "Already staked");
+        }
+
+        _stake(msg.sender, stakeType, duration, uint240(rate), amount);
+    }
+
+    /**
+     * @dev Stake an approved amount of staking token into the contract.
+     *      User must have already approved the contract for specified amount.
+     * @param amount Number of tokens to stake in 1e18
+     * @param duration Number of seconds this stake will be held for
+     */
+    function stake(uint256 amount, uint256 duration) external {
+        require(amount > 0, "Cannot stake 0");
+
+        uint240 rewardRate = _findDurationRate(duration);
+        require(rewardRate > 0, "Invalid duration"); // we couldn't find the rate that corrospond to the passed duration
+
+        // transfer in the token so that we can stake the correct amount
+        stakingToken.safeTransferFrom(msg.sender, address(this), amount);
+        _stake(msg.sender, USER_STAKE_TYPE, duration, rewardRate, amount);
+    }
+
+    /**
+     * @dev Exit out of all possible stakes
+     */
+    function exit() external {
```

Comment:
@micahalcorn This withdraws all the stakes that are passed their lockup period. I assume it's ok from a product standpoint (e.g. we don't need to have the abi=lity to exit just a specific stake).

Reply:
Yes, we talked through that one, thanks @franckc.